src/hotspot/cpu/aarch64/aarch64_sve_ad.m4

```diff
@@ -1664,369 +2001,391 @@ VSHIFT_COUNT(vshiftcntS, H,  8, T_SHORT)
+    __ sve_fcvtzs(as_FloatRegister($dst$$reg), __ S, ptrue, as_FloatRegister($src$$reg), __ S);
+    if (to_bt == T_LONG) {
+      __ sve_vector_extend(as_FloatRegister($dst$$reg), __ D, as_FloatRegister($dst$$reg), __ S);
+    }
   %}
   ins_pipe(pipe_slow);
-%}')dnl
-dnl
-dnl
-define(`VECTOR_CAST_F2X_NARROW2', `
-instruct vcvt$1to$2`'(vReg dst, vReg src, vReg tmp)
+%}
+
+instruct vcvtFtoD(vReg dst, vReg src)
 %{
   predicate(UseSVE > 0 &&
-            n->bottom_type()->is_vect()->element_basic_type() == T_`'TYPE2DATATYPE($2));
-  match(Set dst (VectorCast$1`'2X src));
-  effect(TEMP_DEF dst, TEMP tmp);
-  ins_cost(4 * SVE_COST);
-  format %{ "sve_$3  $dst, $4, $src, $4\n\t"
-            "sve_$5  $tmp, $6, 0\n\t"
-            "sve_$7  $dst, $6, $dst, tmp\n\t"
-            "sve_$7  $dst, $8, $dst, tmp\n\t# convert $1 to $2 vector" %}
+            n->bottom_type()->is_vect()->element_basic_type() == T_DOUBLE);
+  match(Set dst (VectorCastF2X src));
+  ins_cost(2 * SVE_COST);
+  format %{ "sve_vectorcast_f2d  $dst, $dst\t# convert F to D vector" %}
   ins_encode %{
-    __ sve_$3(as_FloatRegister($dst$$reg), __ $4, ptrue, as_FloatRegister($src$$reg), __ $4);
-    __ sve_$5(as_FloatRegister($tmp$$reg), __ $6, 0);
-    __ sve_$7(as_FloatRegister($dst$$reg), __ $6, as_FloatRegister($dst$$reg), as_FloatRegister($tmp$$reg));
-    __ sve_$7(as_FloatRegister($dst$$reg), __ $8, as_FloatRegister($dst$$reg), as_FloatRegister($tmp$$reg));
+    __ sve_vector_extend(as_FloatRegister($dst$$reg), __ D, as_FloatRegister($src$$reg), __ S);
+    __ sve_fcvt(as_FloatRegister($dst$$reg), __ D, ptrue, as_FloatRegister($dst$$reg), __ S);
   %}
   ins_pipe(pipe_slow);
-%}')dnl
-dnl
-dnl
-define(`VECTOR_CAST_F2X_EXTEND1', `
-instruct vcvt$1to$2`'(vReg dst, vReg src)
+%}
+
+instruct vcvtDtoX_narrow(vReg dst, vReg src, vReg tmp)
 %{
   predicate(UseSVE > 0 &&
-            n->bottom_type()->is_vect()->element_basic_type() == T_`'TYPE2DATATYPE($2));
-  match(Set dst (VectorCast$1`'2X src));
-  ins_cost(2 * SVE_COST);
-  format %{ "sve_$3  $dst, $4, $src, $4\n\t"
-            "sve_$5  $dst, $6, $dst\t# convert $1 to $2 vector" %}
+            (n->bottom_type()->is_vect()->element_basic_type() == T_BYTE ||
+             n->bottom_type()->is_vect()->element_basic_type() == T_SHORT ||
+             n->bottom_type()->is_vect()->element_basic_type() == T_INT));
+  match(Set dst (VectorCastD2X src));
+  effect(TEMP_DEF dst, TEMP tmp);
+  ins_cost(3 * SVE_COST);
+  format %{ "sve_vectorcast_d2x  $dst, $src\t# convert D to X vector (narrow)" %}
   ins_encode %{
-    __ sve_$3(as_FloatRegister($dst$$reg), __ $4, ptrue, as_FloatRegister($src$$reg), __ $4);
-    __ sve_$5(as_FloatRegister($dst$$reg), __ $6, as_FloatRegister($dst$$reg));
+    BasicType to_bt = Matcher::vector_element_basic_type(this);
+    Assembler::SIMD_RegVariant to_size = __ elemType_to_regVariant(to_bt);
+    __ sve_fcvtzs(as_FloatRegister($dst$$reg), __ D, ptrue, as_FloatRegister($src$$reg), __ D);
```

Comment:
Converting from double to long and then narrow to target types did not follow JLS. I will fix it. Thanks to @fg1417 for helping to find out this issue.

Reply:
> Converting from double to long and then narrow to target types did not follow JLS. I will fix it. Thanks to @fg1417 for helping to find out this issue.

Fixed in the new commit. Thanks to @PaulSandoz for integrating the fix!

Hi Nick @nick-arm ,

Could you please help to review the new commit, which fixes the same issue as JDK-8276151 for SVE? Thanks!